src/libs/common/types/GeneralizedRange.ts

```diff
@@ -0,0 +1,37 @@
+import { Position } from "./Position";
+import { Range } from "./Range";
+import { TextEditor } from "./TextEditor";
+
+export interface LineRange {
+  type: "line";
+  start: number;
+  /**
+   * Last line, inclusive
+   */
+  end: number;
+}
+
+export interface CharacterRange {
+  type: "character";
+  start: Position;
+  end: Position;
+}
```

Comment:
This is where I landed with the two different range types.  I wouldn't say I'm totally in love with it, but it's cleaner than the other things I tried 🤷